Add imdb example

tommywei110 opened the pull request.

1025KB commented on tfx/examples/imdb/imdb_pipeline_native_keras.py:
```diff
@@ -0,0 +1,196 @@
+
+def _create_pipeline(pipeline_name: Text, pipeline_root: Text, data_root: Text,
+                     module_file: Text, serving_model_dir: Text,
+                     metadata_path: Text) -> pipeline.Pipeline:
```
could you pass in the _beam_pipeline_args like other examples?

tommywei110 replied:
yes! I think i was looking at an old example the last time I updated this

1025KB commented on tfx/examples/imdb/imdb_pipeline_native_keras.py:
```diff
@@ -0,0 +1,196 @@
+    '--direct_running_mode=multi_processing',
+    # 0 means auto-detect based on on the number of CPUs available
+    # during execution time.
+    '--direct_num_workers=1',
```
why change to 1?

tommywei110 replied:
The Bert pipeline has a known error that needs this to be 0, but doesn't concern this example.
Fixed!

1025KB replied:
need to commit again? it still shows 1 here

1025KB commented on tfx/examples/imdb/imdb_pipeline_native_keras.py:
```diff
@@ -0,0 +1,196 @@
+# Lint as: python2, python3
```
please do a side by side comparison for this file vs iris_pipeline_native_keras.py

other than tuner part, it should be about the same.

tommywei110 replied:
fixed

1025KB commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,213 @@
+# Lint as: python2, python3
```
please double check the diff between this file and iris_utils_native_keras.py, other than the tuner part and some function contents (), structure should be about the same

tommywei110 replied:
yes

davidzats-eng commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+from tfx.components.trainer.executor import TrainerFnArgs
+
+_FEATURE_KEY = 'text'
+_LABEL_KEY = "label"
```
be consistent in using single or double quotes throughout the file.

tommywei110 replied:
nice catch. but i think that's the only place in this file. I try to stay consistent and use single quote as much as possible.

davidzats-eng commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _tokenize_review(review):
+  """Tokenize the reviews by spliting the reviews, then constructing a
```
The first sentence in a comment should be no more than one. If you have more to add. Then skip a line after that and add more content after.

tommywei110 replied:
fixed

davidzats-eng commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+  # default value to -1 and add 1 to every index.
+  review_indices = tft.compute_and_apply_vocabulary(
+      review_sparse,
+      default_value=-1,
```
As per earlier comment, it might be best to set default_value to the largest value. Then the model can differentiate between an oov word and the end of a sentence.

tommywei110 replied:
I see your point now. I need to add 1 to max_feature to embedding layer input then.
Need to test if this affect training though.

tommywei110 replied:
*VOCAB_SIZE

davidzats-eng commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+  tokenized = _tokenize_review(inputs[_FEATURE_KEY])
+  return {
+      _transformed_name(_LABEL_KEY): inputs[_LABEL_KEY],
+      _transformed_name(_FEATURE_KEY, True): tokenized}
```
nit: go ahead and replace this with _tokenize_review() and remove the tokenized var.

tommywei110 replied:
sg. fixed.

1025KB commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+_TRAIN_BATCH_SIZE = 10
+
+def _transformed_name(key, is_input=False):
+  return key + ('_xf_input' if is_input else '_xf')
```
just _xf? it's basically just a mark for transformed feature

https://github.com/tensorflow/tfx/blob/master/tfx/examples/iris/iris_utils_native_keras.py#L49

tommywei110 replied:
for some reason, the keras embedding_layer is expecting _input at the end. So this is necessary.

1025KB commented on tfx/examples/imdb/imdb_utils_native_keras.py:
```diff
@@ -0,0 +1,215 @@
+
+  model.fit(
+      train_dataset,
+      epochs=1,
```
this is default if I remember correctly, don't need to set it

tommywei110 replied:
Yes it is.
fixed

tommywei110 commented:
@googlebot I fixed it.

tommywei110 commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Ftfx%2Fpull%2F2049) for more info**.
@googlebot I fixed it.

tommywei110 commented:
@googlebot I fixed it.

liamcrawford commented on tfx/examples/imdb/README.md:
```diff
@@ -0,0 +1,46 @@
+# IMDB Sentiment Analaysis Example
+
+* The data is downloaded from tensorflow_dataset.
+* The example dataset imdb_small.csv that comes within /data is the first 100 entries of the original dataset.
+* To fetch the entire dataset, please use the snippet below 
+* And please adjust the corresponding hyperparameters to account for the larger dataset.
```
Nit: Not sure if these should all be bullets or just a paragraph. Also, if they should be bullets, we should condense that last two into a single one.

In general, this README looks different from the other example READMEs. That said, others have signed off, and I'm satisfied leaving as is if they are.
